proposals/0006-Keyboard-Friendly-Virtualization.md

```diff
@@ -0,0 +1,190 @@
+1. **"focused"** Realize cells around a cell with most recent focus. Can be
+tracked by listening to bubbling onFocusCapture in the default
+`CellRendererComponent`.
+1. **"home"** The area at the beginning of the list (for home key).
+    - Note that part of the top may already reamain realized, since the initial
+    render is retained. This may not be the case with non-default
+    `initialItemIndex` however, and the height may not correspond to layout
+    window.
+1. **"end"** The area at the end of the list (for end key).
+
+> Note that Apple devices have "fn + left" or "fn + right" which behave similarly
+> to home/end, but do not move selection in apps like Finder.
+
+The above scenarios are not exhaustive. It may be required to add additional
+APIs allowing external users greater control over realization behavior. These 
+ould potentially be exposed in the future as dynamic/user-controllable
+realization windows, or separate high-level APIs.
+
+### Batch rendering behavior
+
+After rendering a predefined number of initial items, `VirtualizedList` renders
+new items in batches. Batches may not render the full extent of items to
+realize, and the number of items to render per-batch is configurable.
+
+Cells in realization windows should be automatically rendered, even if not yet
+visible. This can be done during batch renders, giving priority to first render
+the window around visible content.
+
+### Endpoint-specific default behaviors
+
+VirtualizedList should be keyboard-friendly by default on desktop platforms as
+they are oriented towards usage of a physical keyboard. This additional
+realization is not free however, increasing memory usage to retain additional
+cells. This tradeoff may not make sense for mobile.
+
+This leads to desired behavior of:
+- **Keep FlatList focus-safe on desktop:** Realization windows needed for
+correct keyboard navigation should be enabled by default.
+- **Keep FlatList memory usage steady on mobile:** New realization windows
+are opt-in where not pay to play.
+
+The full matrix of behaviors is described in the *API Reference* below.
+
+Endpoint detection is generally performed via `Platform` APIs, such as
+`Platform.isTV()` or `Platform.isPad()`. This change would add a
+`Platform.isDesktop()` endpoint query to allow desktop-specific behavior.
+
+### New VirtualizedList props
+
+Properties should be added to VirtualizedList (and FlatList/SectionList) to
+allow optimizing for app-specific needs. This can be exposed as a
+`RealizationWindowConfig`, describing the size and enablement of predefined
+realization windows. This API could be expanded in the future to allow
+dynamically controlled realization behaviors.
+
+
+### API reference
+
+```tsx
+/**
+ * Allows enabling or disabling an area for realization, optionally overriding
+ * its window size.
+ */
+export type RealizationWindow =
+  | boolean
+  | [boolean, {windowSize?: number}];
```

Comment:
I feel like this might be too much unnecessary configuration being able to adjust each `windowSize`. Any reason we'd ever want this to not be 1?

Reply:
The default `windowSize` for the viewport is fairly large, rendering a whole 10 screens above and below. This is a pretty aggressive amount of pre-rendering. It seems like existing setup optimizes for avoiding blanking on fast movement/interaction over memory consumption.

The realization areas have the potential for similar issues. A conceivable user scenario is to quickly tab between items taller than a viewport. Items would possibly not be realized in time for the keyboard interaction. Keeping a larger window fixes this, but a viewport's worth of native controls could be a non-trivial amount of memory, or cause jank due to excessive native renders.

The right amount of prerendering here seems to depend on scenario. I think we can pick something sane as a default, but am not confident a specific size would work well in all scenarios

I think there might be some alternate APIs to make configuration cleaner, and want to play around with that. I think configurability would be useful here in a general-sense though.